deepchem/models/tensorgraph/models/test_symmetry_functions.py

```diff
@@ -0,0 +1,66 @@
+import scipy
+import numpy as np
+import unittest
+
+from deepchem.models import ANIRegression
+import deepchem as dc
+
+class TestANIRegression(unittest.TestCase):
+
+  def test_gradients(self):
+
+    max_atoms = 3
+
+    X = np.array([
+      [1, 5.0, 3.2, 1.1],
+      [6, 1.0, 3.4, -1.1],
+      [1, 2.3, 3.4, 2.2]
+    ])
+
+    X = X.reshape((1, X.shape[0], X.shape[1]))
+
+    y = np.array([2.0])
+    y = y.reshape((1,1))
+
+    layer_structures = [128, 128, 64]
+    atom_number_cases = [1, 6, 7, 8]
+
+    model = ANIRegression(
+      1,
+      max_atoms,
+      layer_structures=layer_structures,
+      atom_number_cases=atom_number_cases,
+      batch_size=1,
+      learning_rate=0.001,
+      use_queue=False,
+      mode="regression")
+
+    print(X.shape, y.shape)
+
+    train_dataset = dc.data.NumpyDataset(X, y, n_tasks=1)
+
+    model.fit(train_dataset, nb_epoch=2, checkpoint_interval=100)
+
+    new_x = np.array([
+      -2.0, 1.2, 2.1,
+      1.3, -6.4, 3.1,
+      -2.5, 2.4, 5.6,
+    ])
+
+    new_atomic_nums = np.array([1,1,6])
+
+    grad_approx = scipy.optimize.approx_fprime(
+      new_x,
+      model.pred_one,
+      1e-4,
+      new_atomic_nums)
+
+    grad_exact = model.grad_one(new_x, new_atomic_nums)
+
+    print(grad_approx)
+    print(grad_exact)
+
+    np.testing.assert_array_almost_equal(grad_approx, grad_exact)
```

Comment:
@peastman this is the gradient code when comparing finite vs analytic:

AssertionError: 
Arrays are not almost equal to 6 decimals
the assertion raises:
```
(mismatch 100.0%)
 x: array([-0.00596 ,  0.021458,  0.064373,  0.      ,  0.002384,  0.      ,
        0.009537, -0.019073, -0.063181])
 y: array([-0.00846 ,  0.020364,  0.063757, -0.000621,  0.001431, -0.000188,
        0.009081, -0.021795, -0.063569], dtype=float32)
```

Reply:
`approx_fprime()` uses a forward difference, which is only accurate to first order: `(f(x+delta)-f(x))/delta`.  Instead, use a central difference, which is accurate to second order: `(f(x+delta)-f(x-delta))/(2*delta)`.  That should bring down the approximation error a lot.  Once you've done that, your result may be dominated by numerical error, so try increasing the delta to 1e-3 or possibly even 1e-2.  Find the value that gives the best balance of approximation error and numerical error on this problem.

Reply:
Thanks - that's helpful to know. I'll try it out and let you know how it looks.

Reply:
Thanks - that seems to have improved to error to around 3 decimal places, which I can live with for now.